### PR TITLE
[FLINK-28992][table-planner] fix: Change Ndv takes the max value instead of sum of all partitions when getting partition table column stats

### PR DESCRIPTION
**Pull request by swuferhong**
## What is the purpose of the change

Now, when we obtain column statistics `Ndv` of the partition table, we use the method of taking sum of` Ndv` of all partitions, which will cause the obtained stats to be far from the real stats. So now we take the max value instead of sum of all partitions.

## Brief change log

- Change `Ndv` takes the max value instead of sum of all partitions when getting partition table column stats.
- Modify related test.

## Verifying this change

- Modify related test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector:  no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

**Comment by swuferhong** [comment]
@flinkbot run azure

**Comment by swuferhong** [comment]
@flinkbot run azure

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/plan/stats/TableStatsTest.java [hunk 1]
```diff
@@ -55,23 +63,29 @@ void testMergeLackColumnStats() {
...
-        colStatsMerge.put("a", new ColumnStats(7L, 20L, 7D, 23, 35, 2));
-        assertThat(stats1.merge(stats2)).isEqualTo(new TableStats(62, colStatsMerge));
+        colStatsMerge.put("a", new ColumnStats(4L, 20L, 7D, 23, 35, 2));
+        assertThat(stats1.merge(stats2, null)).isEqualTo(new TableStats(62, colStatsMerge));
+        // test column stats merge while column 'a' is partition key. Merged Ndv for columns which
+        // are partition keys using sum instead of max.
+        Map<String, ColumnStats> colStatsMerge2 = new HashMap<>();
+        colStatsMerge2.put("a", new ColumnStats(7L, 20L, 7D, 23, 35, 2));
+        assertThat(stats1.merge(stats2, new HashSet<>(Collections.singletonList("a"))))
+                .isEqualTo(new TableStats(62, colStatsMerge2));
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/plan/stats/TableStatsTest.java:73]: The test does not cover the non-partition columns merge

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/CatalogTableStatisticsConverter.java [hunk 2]
```diff
@@ -80,7 +82,16 @@ public static TableStats convertToAccumulatedTableStates(
...
+
+        if (tableStats.isEmpty()) {
+            return TableStats.UNKNOWN;
+        }
+
+        TableStats resultTableStats = tableStats.get(0);
+        for (int i = 1; i < tableStats.size(); i++) {
+            resultTableStats = resultTableStats.merge(tableStats.get(i), partitionKeys);
+        }
+        return resultTableStats;
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/CatalogTableStatisticsConverter.java:94]: return tableStats.stream().reduce((s1, s2) -> s1.merge(s2, partitionKeys)).orElse(TableStats.UNKNOWN);

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkRecomputeStatisticsProgram.java [hunk 3]
```diff
@@ -207,12 +209,24 @@ private Optional<TableStats> getPartitionStats(
...
         }
     }
 
+    private static Set<String> getPartitionKeys(List<CatalogPartitionSpec> catalogPartitionSpecs) {
+        Set<String> partitionKeys = new HashSet<>();
+        for (CatalogPartitionSpec catalogPartitionSpec : catalogPartitionSpecs) {
+            Map<String, String> partitionSpec = catalogPartitionSpec.getPartitionSpec();
+            for (Map.Entry<String, String> entry : partitionSpec.entrySet()) {
+                partitionKeys.add(entry.getKey());
+            }
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkRecomputeStatisticsProgram.java:225]: partitionKeys.addAll(partitionSpec.keySet());

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java [hunk 4]
```diff
@@ -312,18 +313,25 @@ private static Long getPartitionColumnNullCount(
                                                                     p.getParameters())
                                                             .getRowCount()))
                             .collect(Collectors.toList());
-            Optional<TableStats> optionalTableStats =
-                    catalogTableStatistics.stream().reduce(TableStats::merge);
-            if (!optionalTableStats.isPresent()) {
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java:317]: return tableStats.stream()
                .reduce((s1, s2) -> s1.merge(s2, partitionKeys))
                .orElse(TableStats.UNKNOWN);

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution, I left a few comments

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/CatalogTableStatisticsConverter.java (hunk 2 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/CatalogTableStatisticsConverter.java:94]: > return tableStats.stream().reduce((s1, s2) -> s1.merge(s2, partitionKeys)).orElse(TableStats.UNKNOWN);

Done!

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java (hunk 4 above)
Comment (swuferhong) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java:317]: > return tableStats.stream() .reduce((s1, s2) -> s1.merge(s2, partitionKeys)) .orElse(TableStats.UNKNOWN);

Done!

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkRecomputeStatisticsProgram.java (hunk 3 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkRecomputeStatisticsProgram.java:225]: > partitionKeys.addAll(partitionSpec.keySet());

Done!

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/plan/stats/TableStatsTest.java (hunk 1 above)
Comment (swuferhong) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/plan/stats/TableStatsTest.java:73]: > 

Done!

**Review by godfreyhe — APPROVED** [review]
Thanks for the update, LGTM

**Pull request closed without merging**